### PR TITLE
Finer grained error records for provider load/init failures

### PR DESCRIPTION
**Pull request by levitte**
Finer grained error records for provider load/init failures

When a provider is activated, these three cases would record that the
provider init function failed (implying that it was called):

-   failure to load the provider module (in case it's a dynamically
    loadable module)
-   the init function not being present (i.e. being NULL)
-   the init function being called and returning an error indication
    (i.e. returning a false value)

This is confusing.

Separating the three cases so that they record different errors will
make it easier to determine causes of failure.

### CODE REVIEW

**Comment by levitte** [comment]
I consider this a bug, which is why I'm targetting both master and 3.0.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
Had we had this done earlier, issues like #19368 would have been answered much much quicker.

**Review by mattcaswell — APPROVED** [review]
Approved at least for master - pending the OTC discussion for whether this is ok for 3.0 or not.

**Comment by t8m** [comment]
I agree that this is rather a bug than feature because the consequence is inability to properly diagnose a common failure.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by slontis — COMMENTED** [review]
Could you possibly reword the first sentence of this commit. Your sentence about things being confusing is confusing :)

**Comment by levitte** [comment]
> Could you possibly reword the first sentence of this commit. Your sentence about things being confusing is confusing :)

In a way, that's kinda appropriate, no? 😆 

But yeah ok, I'll give that a go

**Comment by levitte** [comment]
> > Could you possibly reword the first sentence of this commit. Your sentence about things being confusing is confusing :)
> 
> In a way, that's kinda appropriate, no? laughing
> 
> But yeah ok, I'll give that a go

Done.

File: crypto/provider_core.c
Comment (levitte) [reply crypto/provider_core.c]: Fixed

File: crypto/provider_core.c
Comment (levitte) [reply crypto/provider_core.c]: `prov->name` added

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@mattcaswell @beldmit still OK?

**Review by mattcaswell — APPROVED** [review]
Still ok. Subject to the OTC discussion about whether this should go back to 3.0 or not.

**Comment by levitte** [comment]
Merged to master:

2d23ba14630551ee347acafcab81fa1a290c6504 Finer grained error records for provider load/init failures

**Comment by levitte** [comment]
Waiting for OTC approval to cherry-pick 2d23ba14630551ee347acafcab81fa1a290c6504 to 3.0.

**Comment by t8m** [comment]
**OTC: This is a bug fix so it is OK for all branches.**

**Comment by levitte** [comment]
Cheery-picked to 3.0:

840a82982976a0ee4336a3ee0dc4f389aac14b01 Finer grained error records for provider load/init failures

**Comment by levitte** [comment]
Cherry-picked to 3.1:

22b486e2ac07463b29755fb227a913f09bddc5d2 Finer grained error records for provider load/init failures

**Comment by kaend** [comment]
Fails to build with `no-dso`
`/Configure no-dso mingw64 && make`

```
crypto/provider_core.c:912:44: error: 'ERR_R_DSO_LIB' undeclared (first use in this function); did you mean 'ERR_R_DSA_LIB'?
  912 |             ERR_raise_data(ERR_LIB_CRYPTO, ERR_R_DSO_LIB,
      |                                            ^~~~~~~~~~~~~
      |                                            ERR_R_DSA_LIB
```

**Pull request closed without merging**
